{% extends "base_generic.html" %}
{% load static %}

{% block page-title %}<title>HELP ME!!!</title>{% endblock %}

{% block title %}<div class="logo"><h2>Help</h2></div>{% endblock %}

{% block content %}
	<section id="two" class="wrapper style3">
	    <div class="inner">
				<!-- Elements -->
					<div class="row 200%">
						<div class="6u 12u$(medium)">

							<!-- Text stuff -->
							<h3>About this site</h3>
							<h5>ENIGMA genome browser is a genome database designed for comparative analysis of prokaryotic genomes. There are two versions of the genome database: public and internal.</h5>
							<h5>The public version (<a href="https://iseq.lbl.gov/genomes/">https://iseq.lbl.gov/genomes</a>) contains ENIGMA genomes deposited in NCBI GenBank, all genomes from the public version of <a href="http://fit.genomics.lbl.gov/">Fitness Browser</a> and several public genomes of interest.</h5>
							<h5>The internal ENIGMA version (<a href="https://iseq.lbl.gov/genomes-enigma/">https://iseq.lbl.gov/genomes-enigma</a>) of the database contains all genomes from public database and unpublished ENIGMA genomes released for internal use in KBase platform. Access to the ENIGMA internal database requires login and password and is limited to ENIGMA researchers.</h5>
							<h5>All genomes were analyzed with publicly available sequence analysis tools for functional annotation through orthology assignment (EggNOG-mapper) and protein domain mapping (HMMer).</h5>

							<h3>Strains</h3>
							<h5><a href="{% url 'strain_list' %}">List of strains</a> is available from the start page (by click on the number of strains) or from the site menu. Click on the name of a strain in the list to open strain details page. The strain details page contains links to genome assemblies for the strain and strain data (where available).</h5>

							<h3>Genomes</h3>
							<h5><a href="{% url 'genome_list' %}">List of genomes</a> is available from the start page (by click on the number of genomes) or from the site menu. Click on the name of a genome in the list to open genome page. A genome page has links to the list of genes, interactive genome browser and a number of fields for search of genes by functional annotations.</h5>

							<h3>Genes</h3>
							<h5>There are several ways to find a gene of interest. A <a href="{% url 'gene_list' %}">list of all genes</a> (very large) is available from the start page (by click on the number of genes) or from the site menu. You can search genes by name or function from search form at the top of that page. Or, you can open a list of genes for a genome of interest from the genome page. Finally, you can search by functional annotation or by nucleotide or amino acid sequence.</h5>
							<h5>Gene page contains a lot of information. At the top of the page, there are genome name, gene coordinates and gene function that came with the genome file. Next to them, for codnig genes you can find description and taxonomy from EggNOG-mapper output with a list of orthologous groups from EggNOG. Click on the name of an orthologous group opens comparative viewer for that group.</h5>
							<h5>Under the gene information block, you will see interactive genome browser showing that gene (highlighted) and surrounding areas. Click on a gene in the interactive browser opens a pop-up with a link to the gene's page, nucleotide sequence of the gene and some details about the gene.</h5>
							<h5>The next block shows functional annotations for coding genes. Some of them are generated by EggNOG-mapper (like KEGG mappings, EC number, COG class, GO terms etc.). Another annotations may include link to Fitness Browser, Pfam domain or TIGRFAM family mappings.</h5>
							<h5>At the bottom of the page you can find links to external tools for sequence analysis and amino acid sequence (only for coding genes).</h5>
						</div>
						<div class="6u$ 12u$(medium)">
							<h3>Comparative gene view</h3>
							<h5>Click on a name of orthologous group in a gene page opens comparative view for this group. The view shows gene neighborhood around members of the orthologous group. The starting gene is always in the top row, other rows show orthologs from closest organisms in the NCBI taxonomy. If orthologous group has more than 50 members, only 50 genes will be shown.</h5>
							<h5>Please note that the orthologs on this page are not necessarily closest relatives of the starting protein by sequence similarity, since orthologous group may contain several genes from one genome.</h5>
							<h5>Click on a gene opens page for that gene. Genes on the plot are colored according to the orthology assignment by EggNOG-mapper. Coloring by cluster of orthologous genes (COG) is preferred. If the gene was not assigned to any COGs, EggNOG identifier is used for coloring. If a gene was assigned to multiple COGs, a COG with the least number is selected for coloring.</h5>

							<h3>Search in functional annotations</h3>
							<h5>Open <a href="{% url 'textsearch' %}">text search page</a> to find a functional category (like KEGG pathway or EC number) of interest. You can enter an identifier or keywords in the search form to filter list of categories, or click a link next to the search form to open entire list of the categories. Then click on the name of a category to open list of all genes assigned to that category. Functional categories available for search are:</h5>
							<ul>
								<li><h5>KEGG Orthology groups</h5></li>
								<li><h5>KEGG Reactions</h5></li>
								<li><h5>KEGG Pathways</h5></li>
								<li><h5>Enzymatic activities (EC)</h5></li>
								<li><h5>Transporter families (TCDB)</h5></li>
								<li><h5>CAZy families</h5></li>
								<li><h5>COG classes</h5></li>
								<li><h5>GO terms</h5></li>
							</ul>
							<h5>Search in other functional annotations (including Pfam and TIGRFAM) is not supported.</h5>

							<h3>Search by protein or nucleotide sequence</h3>
							<h5>The search application uses MMseqs2 for fast and sensitive sequence similarity search. Open <a href="{% url 'proteinsearch' %}">protein sequence search</a> or <a href="{% url 'nucleotidesearch' %}">nucleotide sequence search</a> page from the start page or from the site menu. Enter your sequence in the FASTA format in the search form and click "Start search". Search with multiple sequence queries is not supported.</h5>
							<h5>Sequence search can take some time depending on the availability of computational resources, so please do not reload the page. Nucleotide sequence search takes more time.</h5>
							<h5>On the protein search results page, a table of hits shows target genes (with link to the gene page), amino acid identity %, length of alignment, E-value and bit-score.</h5>
							<h5>On the nucleotide search results page, a table of hits shows target genomes (with link to the genome page), hit identity %, length of alignment, E-value and bit-score. Click on a genome name opens the genome page with the highlighted hit area in the interactive browser. Congratulations, you might be the second person who read to the very end of the help page.</h5>


						</div>
					</div>
		</div>
	</section>


{% endblock %}
